packages/affiliates/test/mocha/dappmining.js

```diff
@@ -0,0 +1,230 @@
+const Path = require("path");
+const assert = require("assert");
+const highland = require("highland");
+
+const { DappMining } = require("../../libs/affiliates");
+const { getWeb3 } = require("@uma/common");
+const { getAbi } = require("@uma/core");
+const { mocks } = require("../../libs/datasets");
+const { SharedAttributions, Balances } = require("../../libs/models");
+const { EmpAttributions } = require("../../libs/processors");
+const { Queries } = mocks;
+const { encodeAttribution, EncodeCallData } = require("../../libs/contracts");
+
+const web3 = getWeb3();
+const { fromWei, toWei } = web3.utils;
+const empAbi = getAbi("ExpiringMultiParty");
+const datasetPath = Path.join(__dirname, "../datasets/dapp-mining-set1");
+const config = require("../datasets/dapp-mining-set1/config.json");
+
+const encodeCallData = EncodeCallData(empAbi);
+async function makeTx(blockNumber, collateral, tokens, user, tag) {
+  return {
+    blockNumber,
+    name: "create",
+    from_address: user,
+    args: [collateral, tokens],
+    input: encodeAttribution(encodeCallData("create", [collateral.toString()], [tokens.toString()]), tag)
+  };
+}
+
+describe("DappMining", function() {
+  let dappmining, utils;
+  before(function() {
+    const queries = Queries(datasetPath);
+    dappmining = DappMining({ queries, empAbi, web3 });
+    utils = dappmining.utils;
+  });
+  describe("Dappmining Unit Tests", function() {
+    it("calculates percent", function() {
+      let result = utils.calculatePercent(100, toWei(".01"));
+      assert.equal(result, "1");
+      result = utils.calculatePercent(100, toWei("1"));
+      assert.equal(result, "100");
+    });
+    it("sums attributions", function() {
+      const attributions = SharedAttributions();
+      let sum;
+      sum = utils.sumAttributions(attributions);
+      assert(sum);
+      attributions.attribute("a", "a", "1");
+      attributions.attribute("a", "b", "2");
```

Comment:
these attribution calls are essentially saying, attribute creation event by sponsor "a" to developer "b" for 2 tokens. The call is this: `attribution(sponsor,developer,amount)`.

This ends up creating a structure where a user can have attributions by multiple developers (a and b) for different amounts.